### PR TITLE
feat(ERPNext Integrations): Affirm Integration Added

### PR DESCRIPTION
**Pull request by vishdha**
Test Phone Number: 4153334567
Passcode: 1234

Peek 2020-04-13 11-15

### CODE REVIEW

**Comment by Alchez** [comment]
Made upstream to [develop](https://github.com/frappe/erpnext/pull/20968).

**Comment by Alchez** [comment]
@vishdha, check the following notes:

- **Affirm Settings:**
  - [x] Improve the page to separate production and sandbox keys, and remove "Promo Code" and "Pickup Address" and any related logic (the fields have already been changed with https://github.com/DigiThinkIT/erpnext_affirm/pull/18, just remove the logic from address and port it here)
  - [x] Set the different sets of keys mandatory based on whether the "Sandbox Mode" checkbox is selected / unselected.
  - [x] Set both the Private API key fields to a datatype of "Password".

- **Checkout:**
  - [ ] Handle callback redirects from Affirm errors (for example, if you try to check out with a non-US address, Affirm complains with an error. If you click out of the error, the system just stays on that page instead of maybe redirecting back to the cart to retry).
  - [ ] Administrators may not have a linked email address, which is needed to render the Affirm checkout page, see if we can handle that by removing the email requirements, or falling back to a default email address or something.

- **Payments**
  - [x] Hide the "Capture Affirm Payment" in the Sales Order based on the order's billing status.

- **Contribution**
  - [x] Move the "Capture Affirm Payment" into the sales order's controller file (or the place required for regional contributions).
  - [x] Remove all references to "Gateway Selector" from our previous app.
  - [ ] We need to figure out a way to mark this feature for US-only customers. Just confirm if Affirm only offers services for US-based companies, or can any company register for their US customers. We may want to make this a regional feature based on that.

**Comment by DeveloperMetal** [comment]
What are we doing about documentation on this? Noticed it got flagged with "needs-docs"

**Comment by Alchez** [comment]
@forellana-digithinkit, our initial idea was to crowdsource the docs from the community, but as a first draft, we could use their [previous payment gateway docs](https://docs.erpnext.com/docs/user/manual/en/erpnext_integration/contents#payment-integration) as templates, and make our own.

@vishdha or I can probably pick that up after this is ready.

File: erpnext/erpnext_integrations/doctype/affirm_settings/affirm_settings.js [hunk 1]
```diff
@@ -2,7 +2,10 @@
...
 
 frappe.ui.form.on('Affirm Settings', {
-	// refresh: function(frm) {
-
-	// }
+	is_sandbox: function(frm) {
+		frm.toggle_reqd("public_sandbox_api_key", frm.doc.is_sandbox? 1: 0);
+		frm.toggle_reqd("private_sandbox_api_key", frm.doc.is_sandbox? 1: 0);
+		frm.toggle_reqd("public_api_key", frm.doc.is_sandbox? 0: 1);
+		frm.toggle_reqd("private_api_key", frm.doc.is_sandbox? 0: 1);
```
Comment (Alchez) [inline erpnext/erpnext_integrations/doctype/affirm_settings/affirm_settings.js]:
```suggestion
		frm.toggle_reqd("public_sandbox_api_key", frm.doc.is_sandbox);
		frm.toggle_reqd("private_sandbox_api_key", frm.doc.is_sandbox);
		frm.toggle_reqd("public_api_key", !frm.doc.is_sandbox);
		frm.toggle_reqd("private_api_key", !frm.doc.is_sandbox);
```

**Review by DeveloperMetal — APPROVED** [review]
Lets make a separate task for the documentation.
Need to merge this for tomorrow's demo

**Pull request merged**